src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5439,14 +5439,26 @@ private void ParseTypeArgumentList(out SyntaxToken open, SeparatedSyntaxListBuil
             // remaining types & commas
             while (true)
             {
+
                 if (this.CurrentToken.Kind == SyntaxKind.GreaterThanToken)
                 {
                     break;
                 }
-                else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleType())
+                else if (this.CurrentToken.Kind == SyntaxKind.CommaToken)
                 {
-                    types.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
-                    types.Add(this.ParseTypeArgument());
+                    eatCommaAndTypeArgument();
+                }
+                else if (this.IsPossibleType())
+                {
+                    var nextToken = this.PeekToken(1);
```

Comment:
i don't understand how peeking 1 token is sufficient.  the 'possible type' may be unbounded in length.  are you only testing simple `1-token types`.

Reply:
@CyrusNajmabadi 
You are right.
If lookahead is very quick, how far ahead is it performant to lookahead?

I think a sensible algorithm might be to say that a GenericTypeArgumentList will always appear somewhere before one of:
```
;
=
{
=>
```

We look ahead till the next such token, and if we are able to find a closing `>` token (counting opening `<` along the way) we assume we are missing a comma.

Else we assume we are missing a `>` token.

If we get past some contant number of tokens, we give up and assume it's a `,`

Does that sound sensible?

At the cost of making the function longer (but since it's a switch, presumably not much less performant) we could add all the miriad SytntaxKinds that can't appear in a Generic Type Argument List, but I think these syntax kinds will be sufficient.